### PR TITLE
[LMLayer] The ModelCompositor

### PR DESCRIPTION
**Pull request by jahorton**
This PR implements the rest of #1802, adding a new middle layer to the LMLayer called the `ModelCompositor`.  This layer will allow the LMLayer to combine probabilities from KMW's fat-finger analysis code with those from the lexical model (once they are made available), allowing us to, for the first time, support predictive suggestions.

Of course, there are some pretty significant limitations at this stage:

- Corrections will only consider the most recently-typed keystroke as erroneous - it doesn't model correction for multiple typos at this time.
- The probability combination code doesn't yet consider the (hopefully) edge-case where two different keystrokes produce the same resulting output in a manner that allows merging their probability masses.
- The existing model templates were forced into compliance with the new structure, but they don't all yet return meaningful / proper probability masses.
    - Further investigation is needed to return proper probability masses based on word weights for the weighted wordlist model.
- The composited probability distribution may need tweaking.
    - Is `p(word | language) * p(keystroke)` ideal?  Mathematically speaking, they're not independent probabilities... but we are getting these from two very different, independent sources with necessary factorization.

Those caveats aside, fat-finger use is now testable!

One more TODO:  this will break the 'custom' model in our `lexical-models` repo, as we now need models to return `Distribution<Suggestion>` rather than `Suggestion[]`.

### CODE REVIEW

**Comment by keyman-server** [comment]
[OSK renders for Windows](https://build.palaso.org/repository/download/KeymanDesktop_TestPrRenderOnScreenKeyboards/152482:id/kvk/index.html)

File: common/predictive-text/message.d.ts [hunk 1]
```diff
@@ -179,6 +179,26 @@ interface Transform {
...
 }
 
+/**
+ * Represents members of a probability distribution over potential outputs
+ * from ambiguous text sequences.  Designed for use with fat-finger correction
+ * and similar typing ambiguities.
+ */
+interface ProbabilityMass<T> {
+  /**
+   * A potentially-valid Transform usable to produce the next state of input.
```
Comment (eddieantonio) [inline common/predictive-text/message.d.ts]: Ditch the word "Transform", since a `Suggestion` can be used here too.

File: common/predictive-text/worker/models/dummy-model.ts [hunk 2]
```diff
@@ -52,11 +52,29 @@ namespace models {
...
+      let makeUniformDistribution = function(suggestions: Suggestion[]): Distribution<Suggestion> {
+        let distribution: Distribution<Suggestion> = [];
+        let n = suggestions.length;
+
+        for(let s of suggestions) {
+          distribution.push({sample: s, p: 1});  // For a dummy model, this is sufficient.  The uniformness is all that matters.
+        }
+
+        return distribution;
+      }
```
Comment (eddieantonio) [inline common/predictive-text/worker/models/dummy-model.ts]: To not break backwards-compatibility, I suggest letting _all_ models that do not implement `predictWithDistribution()` to return a uniform distribution by default.

File: common/predictive-text/worker/models/dummy-model.ts [hunk 3]
```diff
@@ -52,11 +52,29 @@ namespace models {
...
+
       if (injectedSuggestions) {
-        return injectedSuggestions;
+        return makeUniformDistribution(injectedSuggestions);
+      }
+
+      let currentSet = this._futureSuggestions.shift();
+      
+      if(!currentSet) {
+        return null;
```
Comment (eddieantonio) [inline common/predictive-text/worker/models/dummy-model.ts]: Does this type-check? Should this not return an empty distribution, e.g., `[]` or one with all events having a probability of 0?

File: common/predictive-text/worker/models/trie-model.ts [hunk 4]
```diff
@@ -72,16 +72,28 @@
       };
     }
 
-    predict(transform: Transform, context: Context): Suggestion[] {
+    predict(transform: Transform, context: Context): Distribution<Suggestion> {
+      // TODO:  Get a better distribution!  Our words do have frequency weighting, right?
```
Comment (eddieantonio) [inline common/predictive-text/worker/models/trie-model.ts]: Yes, they do!

We can implement `predictWithDistribution` in a future PR!

File: common/predictive-text/worker/worker-interfaces.ts [hunk 5]
```diff
@@ -118,7 +118,7 @@ interface LMLayerWorkerState {
  */
 interface WorkerInternalModel {
   configure(capabilities: Capabilities): Configuration;
-  predict(transform: Transform, context: Context): Suggestion[];
+  predict(transform: Transform, context: Context): Distribution<Suggestion>;
```
Comment (eddieantonio) [inline common/predictive-text/worker/worker-interfaces.ts]: Idea: do not change the signature of `predict()`; instead, add `predictWithDistribution()` that has this same method signature.

Then move all of the code in `predict()` of existing models in this PR to `predictWithDistribution()`. `predict()` will delegate to `predictWithDistribution()` and will "unwrap" the distribution: essentially all of that `.map(function(value) { return value.sample })` stuff.

File: web/source/text/prediction/modelManager.ts [hunk 6]
```diff
@@ -313,9 +313,7 @@ namespace com.keyman.text.prediction {
       this.recordTranscription(transcription);
 
       let transform = transcription.transform;
-      // TODO:  Use this instead of the single Transform; requires LMLayer work.
-      let alternates = transcription.alternates;
-      var promise = this.currentPromise = this.lmEngine.predict(transform, context);
+      var promise = this.currentPromise = this.lmEngine.predict(transcription.alternates || transcription.transform, context);
```
Comment (eddieantonio) [inline web/source/text/prediction/modelManager.ts]: 👍

**Comment by jahorton** [comment]
> I like the new model compositor, but I'm afraid it makes too many breaking changes.

It's pretty clear that we need a `predict()` that provides a probability distribution... which is why I want to make the change now.

> Then, `predict()` would have its old method signature and behaviour, but, I guess we'd have to deprecate it :/

Let me get this straight... we should deprecate a method that's _only ever existed in alpha_ before we even open up the beta?  We'd (seemingly) be forcing model authors to implement an already-deprecated method that we already don't _really_ want to support.

We've noted that now's the time to finalize the lexical model spec, and I think it's better that we go ahead and rip this band-aid off now while it only affects us and our coding, before it also affects anyone else.

File: common/predictive-text/unit_tests/headless/worker-predict-dummy.js [hunk 7]
```diff
@@ -60,7 +60,9 @@ describe('LMLayerWorker dummy model', function() {
         left: "I'm a little ",
         startOfBuffer: true,
         endOfBuffer: true,
-      }, expectedSuggestions);
+      }, expectedSuggestions).map(function(value) {
+        return value.sample;
+      });
```
[the remark this answers is not in the document]
Comment (jahorton) [reply common/predictive-text/unit_tests/headless/worker-predict-dummy.js]: Right, I did forget about `lmlayer-cli`.  If I take the time to adjust it to match this PR and tweak anything affected in the `lexical-models` repo, would that be fine?

Keep in mind, we're only talking about backward-compatibility within this alpha.  There are no prior betas, let alone stables, to support or worry about breaking with these changes.

**Comment by jahorton** [comment]
Testing `lmlayer-cli` with this branch with no further changes:
- the test model "simple-en-wordlist.js" had no problems.
- From the `lexical-models` repo:
    - example.en.wordlist:  no issues
    - example.en.custom:  nothing displayed, but there are no crashes.  It's been in need of updating anyway, so I don't think there's any real loss here.
        - Same behavior on `master`.
    - example.crk.wordlist_wahkohtowin: no issues
    - example.str.trie_sencoten:  I can get suggestions to appear for a null context, but none appear when I type a character.
        - I get identical behavior on `master` though, so... I'm guessing it's related to the keycodes/underlying Unicode somehow.  I don't know Sencoten well enough to really say.

File: common/predictive-text/unit_tests/headless/worker-predict-dummy.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (jahorton) [reply common/predictive-text/unit_tests/headless/worker-predict-dummy.js]: > I believe not only lexical-models will break, but so will the `lmlayer-cli`.

You might be in for a pleasant surprise, btw... see my comment down below:  https://github.com/keymanapp/keyman/pull/1803#issuecomment-496334571

tl;dr:  No breakage, except maybe with en.custom.

File: common/predictive-text/worker/models/dummy-model.ts (hunk 2 above)
Comment (jahorton) [reply common/predictive-text/worker/models/dummy-model.ts]: Yep.  That's why I already implemented this within all existing model templates within this PR:

* `DummyModel`
* `WordListModel`
* `TrieModel`  (in this case, as a stop-gap before returning a weight-based probability instead)

**Comment by eddieantonio** [comment]
>     * I get identical behavior on `master` though, so... I'm guessing it's related to the keycodes/underlying Unicode somehow.  I don't know Sencoten well enough to really say.

Ah yes! This is because the current Sencoten model in lexical-models has the incorrect search key function. I will make a PR, but I was waiting for permission from the Saanich school board to post the FULL word list, in addition to the correct key function.

**Comment by eddieantonio** [comment]
> We've noted that now's the time to finalize the lexical model spec, and I think it's better that we go ahead and rip this band-aid off now while it only affects us and our coding, before it also affects anyone else.

Ah, okay. I guess I just wasn't expecting a huge change like this to happen so late into alpha. Yeah, we can change the interfaces, if you have done the work and changed all sites that use it.

File: common/predictive-text/worker/models/dummy-model.ts (hunk 3 above)
Comment (eddieantonio) [inline common/predictive-text/worker/models/dummy-model.ts]:
```suggestion
        return [];
```

Comment (jahorton) [reply common/predictive-text/worker/models/dummy-model.ts]: Technically, neither are perfectly accurate - the original return statement was as follows:

` return this._futureSuggestions.shift();`

When `_futureSuggestions` is an empty array (`[]`), the `.shift()` will technically return `undefined`.

Or are we saying that we want to "throw this in" as a quick additional enhancement?

Comment (eddieantonio) [reply common/predictive-text/worker/models/dummy-model.ts]: Oh, I neglected to realize the context! Since this is the _dummy_ model, I guess it doesn't super make a difference. Either throw an error (should only be thrown in tests), or an empty array — does it make sense for a model to return 0 suggestions?

Handle it by whatever means makes the most sense. Returning `null` or `undefined` is usually not useful!
